### ISSUE TITLE
opt: extract constraints from timezone(ts, 'Europe/Berlin') >= '2020-..'

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/cockroach#48149]
We need to be able to create index constraints from operations on `WHERE ts AT TIME ZONE 'foo'` which is an alias for `timezone(ts, 'foo')`.

The `timezone` function converts from `DTimestampTZ` to `DTimestamp` by adding the zone offset. One nice solution would be to create the reverse function that converts back, and add a normalization rule that flips `timezone(ts, 'foo') >= 'bar'` to `ts >= from_timezone('bar', 'foo')`.

Pasting @korpa's example from https://github.com/cockroachdb/cockroach/issues/18836#issuecomment-621171675:

```sql
CREATE TABLE table1 (id TEXT, ts TIMESTAMPTZ, kpi1 INT, kpi2 INT, PRIMARY KEY (id));

CREATE INDEX ON table1 (ts) STORING (kpi1,kpi2);
```

without "AT TIME ZONE":

```sql
EXPLAIN
WITH
days AS (
  SELECT generate_series( '2020-04-01T00:00:00', '2020-04-30T23:59:59', interval '1' day)::date  as "dateday"
),
kpis AS (
  SELECT
    date_trunc('day', "ts")::date AS "ts",
    sum("kpi1") as kpi1,
    count("kpi2") as kpi2
  FROM "table1"
  WHERE "ts" BETWEEN '2020-04-01 00:00:00' and '2020-04-30 23:59:59'
  GROUP BY 1
)
SELECT
  days.dateday::DATE as "date",
  COALESCE( kpis.kpi1, 0 ) as "kpi1",
  COALESCE( kpis.kpi2, 0 ) as "kpi2"
FROM days LEFT JOIN kpis ON days.dateday = kpis.ts
ORDER BY days.dateday DESC;
```
![image](https://user-images.githubusercontent.com/2982785/80595750-8c455b00-8a25-11ea-93a3-d0e7e3c590a0.png)

with "AT TIME ZONE"
```sql
EXPLAIN
WITH
days AS (
  SELECT generate_series( '2020-04-01T00:00:00', '2020-04-30T23:59:59', interval '1' day)::date  as "dateday"
),
kpis AS (
  SELECT
    date_trunc('day', "ts" AT TIME ZONE 'Europe/Berlin')::date AS "ts",
    sum("kpi1") as kpi1,
    count("kpi2") as kpi2
  FROM "table1"
  WHERE "ts" AT TIME ZONE 'Europe/Berlin' BETWEEN '2020-04-01 00:00:00' and '2020-04-30 23:59:59'
  GROUP BY 1
)
SELECT
  days.dateday::DATE as "date",
  COALESCE( kpis.kpi1, 0 ) as "kpi1",
  COALESCE( kpis.kpi2, 0 ) as "kpi2"
FROM days LEFT JOIN kpis ON days.dateday = kpis.ts
ORDER BY days.dateday DESC;
```

![image](https://user-images.githubusercontent.com/2982785/80595878-bc8cf980-8a25-11ea-9322-a318072d0a36.png)

```txt
Build Tag:    v20.1.0-rc.2
Build Time:   2020/04/17 19:22:16
Distribution: CCL
Platform:     linux amd64 (x86_64-unknown-linux-gnu)
Go Version:   go1.13.9
C Compiler:   gcc 6.3.0
Build SHA-1:  c0f00d2a156295a1cb79ed960940c167cb03ccbc
Build Type:   release
```

**Comment by mgartner** [comment]
@RaduBerinde Got very close to a solution here, but ran into a last-mile hurdle of `fold_constants.go` (and the check in `norm_test.go`) not allowing folding of "Date and time" functions. This folding is necessary for us to generate constraints for an index scan in the example above.

However, now that #26582 is closed (Thanks @otan, it was incredible timely of you to tackle that!), I can remove this white list: https://github.com/cockroachdb/cockroach/blob/d3aa53d44b528d4f0d120654c709c000222c0a71/pkg/sql/opt/norm/fold_constants.go#L360-L361

Once I get that whitelist removed, I should have a fix for this issue ready to go.

**Comment by RaduBerinde** [comment]
We are not ready to remove that whitelist yet.. The fix for 26582 is for now mainly cosmetic, the optimizer does not yet take into account the new volatility value. I am planning to take a look at that sometime in this release.

Is it sufficient to add the new function to the whitelist? If yes, I'd go ahead and do that. If not, we should merge the new function and the rule for now and we can close the issue later, when the other volatility stuff is in place.

**Comment by mgartner** [comment]
Ya I can do that. I'll just have to remove the extra check in `norm_test.go` that ensures we don't whitelist a date and time function. I don't think removing it will be much of a risk, as long as we are diligent when reviewing changes to the whitelist.

Also, a new function is not needed. `timezone(zone, TIMESTAMPTZ)` is the inverse of `timezone(zone, TIMESTAMP)`.

In Postgres:

```
timezone=# SET TIMEZONE = 'America/New_York';
SET
timezone=# CREATE TABLE t (tz TIMESTAMPTZ);
CREATE TABLE
timezone=# INSERT INTO t VALUES ('2020-06-01 12:35:55-07');
INSERT 0 1
timezone=# SELECT tz = timezone('America/Denver', timezone('America/Denver', tz)) FROM t;
 ?column?
----------
 t
(1 row)
timezone=# SET TIMEZONE = 'Europe/London';
SET
timezone=# SELECT tz = timezone('America/Denver', timezone('America/Denver', tz)) FROM t;
 ?column?
----------
 t
(1 row)
```

And [this blog post](https://phili.pe/posts/timestamps-and-time-zones-in-postgresql/#converting-between-timezones) is a good reference that explicitly calls out that they are inverses.

> This means that when we converted a timestamptz we got back a timestamp, whereas when we converted a timestamp, it gave us a timestamptz. In other words, in the first example we converted an absolute time to a wall time whereas in the second example we did the _**inverse**_ operation, i.e. we converted a wall time back to an absolute time.
> 
> In reality, there are multiple timezone functions, one that takes timestamp and one that takes timestamptz2. I like to think of them as follows:
> 
> * When passing a time zone x and a value t of type timestamptz, i.e., timezone(x, t), I’m basically saying, _give me the time the clock was showing in time zone x at the absolute point in time t_.
> 
> * When passing a time zone x and a value t of type timestamp, I’m saying, _give me the absolute time for when clocks in time zone x were showing t_.

**Comment by RaduBerinde** [comment]
Oh, great find! Yeah I see the implementation of the inverse now, I don't know how I missed it.

I looked at the code more - I don't think is's safe to whitelist the folding of this function in general. The timestamp can refer to the current time, e.g. `select timezone('Europe/Berlin', 'now')`, or `timestamp('Berlin/Europe', timestamptz 'now')`. We have a separate issue for folding `now()` (#18836). I also realized that we currently incorrectly cache the result of parsing `timestamp 'now'` in plans.. filing an issue for that.

I'd go ahead with just the norm rule for now, we have to think about folding time stuff more holistically.

**Comment by mgartner** [comment]
Is `'now'` the only case where `timezone()` is non-immutable?

If so, could we normalize `'now'` within `timezone` (or other functions/casts) to `now()`. I _think_ that would allow us to whitelist `timezone`, because `FoldFunction` won't apply when the function's arguments aren't constants.

**Comment by RaduBerinde** [comment]
There are other cases (things like `'tomorrow'`). We will need to augment the functions that parse timestamps to return information about whether we had a relative-to-now component. And we will need to use this information when folding casts (`timestamp 'now'` is actually syntactic sugar for `'now'::timestamp`) and when folding `timezone()`, and probably other places.

This will need to wait until we have better handling of immutable vs stable values (#18836).

**Comment by korpa** [comment]
Wow. That was fast. Thank you very much. Great team!

Will this be released in v20.1.x or do I have to wait for v20.2?

**Comment by RaduBerinde** [comment]
@mgartner let's try to backport, I don't think it depends on any changes since v20.1. If you haven't done it before, `go get` this utility: https://github.com/benesch/backport

**Comment by mgartner** [comment]
@RaduBerinde SGTM. It's my first backport, so thanks for the link. If I have any questions I'll let you know.

**Comment by RaduBerinde** [comment]
@korpa we've been hitting some issues backporting the fix (it is interacting with some things that are in flux for v20.2). How important is having this fix in v20.1.x for your usecase? Would it be possible to work around it until v20.2 by doing the timezone calculation in the application?

**Comment by korpa** [comment]
@RaduBerinde I already have a workaround in place. I just wanted to get rid of it as soon as possible to clean up my code. It's ok for me if it is released with v20.2. Nevertheless thank you very much for your efforts.
